packages/nimble-components/src/table/specs/table-column-interaction-events.md

```diff
@@ -0,0 +1,60 @@
+# Table Column Interaction Events HLD
+
+## Problem Statement
+
+Some clients need to know when a user interactively changes the configuration of their table columns. This includes:
+
+-   Sorting a column
+-   Grouping by a column
+-   Resizing a column
+
+### Out of scope of this HLD
+
+There are some interactions on the table that will not be covered by this HLD. The rationale for that decision is that they are not interactions that affect a column state. These include:
+
+-   Expanding/collapsing a group
+-   Modifying row selection; row selection can already be observed by a client using the `selection-change` event on the table
+
+## Implementation / Design
+
+### API
+
+In order to keep clients from having to listen for a number of different events when they will likely respond in a similar way to any part of the column configuration changing, there will be a single `column-configuration-change` event that fires when a user interactively sorts a column, groups by a column, or resizes a column. The details of the event will include the current column configuration and look as follows.
+
+```ts
+/**
+ * Event detail type for interactive column configuration changes.
+ */
+export interface TableColumnConfigurationChangeEventDetail {
+    columns: TableColumnConfiguration[];
+}
+
+/**
+ * Information about the current configuration of a single column
+ */
+export interface TableColumnConfiguration {
+    columnId?: string;
+    sortIndex?: number | null;
+    sortDirection: TableColumnSortDirection;
+    groupIndex?: number | null;
+    hidden: boolean;
+    fractionalWidth: number;
+    pixelWidth?: number | null;
+}
+```
+
+Columns will be included in the `columns` array in the same order they are specified in the DOM. Information included in the `TableColumnConfiguration` object will be specified by the "current" values when available (e.g. `currentSortIndex`, `currentSortDirection`). When an attribute does not have a "current" value, then the non-current value will be specified (e.g. `columnId`, `groupIndex`).
+
+## Future Work
```

Comment:
Just a side thought / rambling (no action needed)

Wonder how we would handle the case of column authors wanting to expose configuration settings for a column. This would be outside the systemlink case which delegates most config to a slide out. Or maybe that would just be the recommendation for all NI apps. But thinking about it from the let me render this column as a string / number / url discussion.

Like how perspectives table lets columns plugin config for rendering the cell of a numeric column:
![image](https://github.com/ni/nimble/assets/1588923/112f5a52-cdb2-47e9-8083-0e8b3c14b931)

Maybe we would let them add entries to the column header menu or something.

From that was thinking that maybe they could plugin to column-configuration-changed but actually they should probably just fire events off their column element instance. Our events are just for the state the table is managing. Column authors can make their own events for any custom state they are managing. We maybe would give a hook into the menu in the future (could just forward a slotted menu item action menu style)

Reply:
That's an interesting question and part of the reason why I wanted to punt on get/set column configuration functions until we had more requirements. It's definitely something we'll have to put some more thought into if/when we need to support it.